### ISSUE TITLE
trackedNodes for split edge queries returned in random order causing issues re-building edge fields

### ISSUE THREAD

**Issue by HurricaneJames** [facebook/relay#478]
Summary: Relay does not merge query fragments when calling setVariables. Instead it uses a unique id for each query. This works fine, until a mutation means everything must be refetched. Then, it tries to rebuild edges from those previous queries, but the edges are sorted in alpha order on the unique ids. `putRangeEdges` tosses out a warning (cannot find segment) when it reaches a segment that has a cursor that has not been processed yet, then quietly abandons the data and keeps going.

Full Description:
Lets say we have a simple list component.

````
class ASimpleList extends React.Component {
  onClick = (e) => {
    e.preventDefault();
    this.props.relay.setVariables({
      count: this.props.relay.variables.count + 4
    });
  }
  render() {
    return (
      <div>
        <ul>
        {
          this.props.viewer.listItems.edges.map((itemEdge) =>
            <li key={itemEdge.node.id}>
              <ItemView item={itemEdge.node} />
            </li>
          )
        }
        </ul>
        {
          this.props.viewer.listItems.pageInfo.hasNextPage &&
          <button onClick={this.onClick}>Load More</button>
        }
      </div>
    );
  }
}
````

This Simple component is wrapped in a Relay container.

````
export default Relay.createContainer(ASimpleList, {
  initialVariables: { count: 2 },
  fragments: {
    viewer: () => Relay.QL`
      fragment on RelayViewer {
        id
        listItems(first: $count) {
          edges {
            node {
              id
              ${ItemView.getFragment('item')},
            }
          }
          pageInfo { hasNextPage }
        }
      }
    `,
  }
})
````

Clicking on the "Load More" button works as expected and loads more items. Each time it generates an optimized query fragment to only load the extra. The first fetch, `items0000`, grabs 2 items. The second fetch, `items0002`, only grabs the next 4. The third fetch, `items0001`, requests the next 4 and returns 2 from my sample data.

The fragment names are not the actual names used. Relay uses random keys, `itemsemfrel`, `itemsywxi0w`, and `itemsx3ao8b`. These keys, in an alpha sort may as well be items0000, items0002, and items0001.

Now, lets assume there is another component on the page somewhere else that calls a mutation.

````
import Relay from 'react-relay';

export default class MoveItemMutation extends Relay.Mutation {
  getMutation()     { return Relay.QL`mutation { moveItem }`; }
  getVariables()    {
    return {
      itemId:  this.props.itemId,
      regionId: this.props.targetRegion.id
    };
  }
  getCollisionKey() { return 'moveItem_' + this.props.itemId; }
  getFatQuery()     {
    // return the whole view since moving an item can technically change thing in the system
    return Relay.QL`
      fragment on MoveItemPayload {
        viewer
      }
    `;
  }
  getConfigs() {
    return [
      {
        type: 'FIELDS_CHANGE',
        fieldIDs: {
          viewer: this.props.viewer.id
        }
      },
    ]
  }
}
````

This mutation works fine where it is called. This is because the regions component specifies a query to MAX_INT for region edges and the item edges in those regions. That means there is only a single query to consolidate.

However, because of the load more functionality, the ItemList container ends up with various fragments, which may or may not be ordered correctly when being processed by `RelayRecordStore::putRangeEdges`. This leads to a warning `Warning: GraphQLRange cannot find a segment that has the cursor: [cursorId]`. It continues processing the remaining data, but misses everything from the skipped edges, and I would imagine anything that relies on those edges cursors.

I do not have a solution to this problem yet. It could be that the viewer design an anti-pattern for Relay. In which case, it should probably be avoided in the examples. I suspect merging the queries is impossible (each fragment could have different params, though this seems like Relay simply trying to be too flexible). It might also be possible to use a timestamp in the query fragment name instead of the (random?) characters.

Any suggestions?

**Comment by HurricaneJames** [comment]
The original name for this issue was the console warning. However, that was not very descriptive. The new name helps to clarify the real issue.

After working with the relay code base for a while tonight it looks like any time `props.relay.setVariables` updates an edge fetch count (ex. first: 1 -> first: 5), it splits the query into 2 tracked nodes, `{ first: 1}` and `{ first: 4, after: [current end cursor] }`. The fragments for this split query are run through a global id generator to get their `__serializationKey__`.

This works fine until something, a mutation perhaps, means Relay needs to refetch everything. At which point `RelayQueryTracker::getTrackedChildrenForID` returns a query with the split query fragments ordered on `__serializationKey__`. I believe this is happening in `flattenRelayQuery`.

**Comment by HurricaneJames** [comment]
Not sure why it submitted that last comment, I was not done writing and definitely did not hit comment. 

Anyway, I wanted to point out that it should not matter what order those fragments are in when the query is made because there is really no guarantee that they will come back in that order. Even if they do, it seems all too easy to sort them by accident at some processing step. Instead, it seems like connection queries need to have a sequence value that the GraphQL server would parrot.

**Comment by josephsavona** [comment]
@HurricaneJames this appears to be basically the same issue as #479 as both ultimately are caused by the building of tracked queries for a connection. I'm going to close this - let's move discussion to that issue.

**Comment by HurricaneJames** [comment]
I think this is a different issue. One that actually seems to have a simple fix (even if it did take hours to find and tons of dead ends while I learned the relay code base).

`toQuery` in `flattenRelayQuery.js` sorts the keys. Changing `var keys = Object.keys(flattenedFieldMap).sort(sortTypeFirst);`) to `var keys = Object.keys(flattenedFieldMap)` seems to fix the bug. It also does not seem to make any tests fail. Is there a reason the keys were sorted?

**Comment by wincent** [comment]
> Is there a reason the keys were sorted?

@HurricaneJames: I replied on your PR.

**Comment by josephsavona** [comment]
Fixed in 389d25a
